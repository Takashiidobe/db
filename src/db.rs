use std::{
    collections::BTreeMap,
    fs::{File, OpenOptions},
    io::{BufWriter, Seek as _, SeekFrom, Write as _},
    num::NonZeroU32,
};

use crate::{
    row::{RowType, RowVal, Schema},
    wal::WAL,
};

use crate::page::{Page, PageHeader, PAGE_SIZE};
use indexset::{BTreeSet, Range};

#[derive(Debug)]
pub struct DB {
    pub pages: BTreeSet<(Page, Option<usize>)>,
    pub file: File,
    pub wal: WAL,
    pub epoch: u64,
    pub schema: Schema,
}

impl DB {
    pub fn new(file_name: &str, schema: &[RowType]) -> Self {
        let epoch = 1;
        let (db_file, wal_file, schema_file) = Self::setup_files(file_name, epoch);
        Self {
            file: db_file,
            pages: BTreeSet::new(),
            wal: WAL {
                file: wal_file,
                records: BTreeMap::new(),
            },
            epoch,
            schema: Schema {
                schema: schema.to_vec(),
                file: schema_file,
            },
        }
    }

    pub fn new_with_pages(
        pages: BTreeSet<(Page, Option<usize>)>,
        file_name: &str,
        schema: &[RowType],
    ) -> Self {
        let epoch = 1;
        let (db_file, wal_file, schema_file) = Self::setup_files(file_name, epoch);

        Self {
            file: db_file,
            pages,
            wal: WAL {
                file: wal_file,
                records: BTreeMap::new(),
            },
            epoch,
            schema: Schema {
                schema: schema.to_vec(),
                file: schema_file,
            },
        }
    }

    fn setup_files(file_name: &str, epoch: u64) -> (File, File, File) {
        let db_file = OpenOptions::new()
            .create(true)
            .read(true)
            .write(true)
            .open(format!("{file_name}.{epoch}.db"))
            .unwrap();
        let wal_file = OpenOptions::new()
            .create(true)
            .read(true)
            .append(true)
            .open(format!("{file_name}.{epoch}.wal"))
            .unwrap();
        let schema_file = OpenOptions::new()
            .create(true)
            .read(true)
            .write(true)
            .open(format!("{file_name}.{epoch}.schema"))
            .unwrap();
        (db_file, wal_file, schema_file)
    }

    pub fn sync(&mut self) -> bool {
        // apply all updates in wal to pages
        for (id, val) in self.wal.records.clone() {
            self.insert_to_page(id, &val);
        }

        self.serialize();
        self.wal.records.clear();
        self.wal.file.set_len(0).is_ok()
    }

    pub fn serialize(&self) {
        let mut f = BufWriter::new(&self.file);
        for (i, page) in self.pages.iter().enumerate() {
            if page.0.dirty || page.1 != Some(i) {
                let pos = SeekFrom::Start((i * PAGE_SIZE) as u64);
                let _ = f.seek(pos);
                let _ = f.write_all(&page.0.to_page_bytes());
            }
        }
        // truncation is required otherwise the page might have stale pages that have been deleted.
        let _ = self.file.set_len((self.pages.len() * PAGE_SIZE) as u64);
    }

    fn range_iter(&self, id: NonZeroU32) -> Range<(Page, Option<usize>)> {
        self.pages.range(
            (
                Page {
                    header: PageHeader {
                        end: id,
                        start: NonZeroU32::MIN,
                        count: u32::MIN,
                    },
                    dirty: false,
                    data: BTreeMap::new(),
                    size: 0,
                    schema: vec![],
                },
                None,
            )
                ..=(
                    Page {
                        header: PageHeader {
                            end: NonZeroU32::MAX,
                            start: id,
                            count: u32::MAX,
                        },
                        dirty: true,
                        data: BTreeMap::new(),
                        size: usize::MAX,
                        schema: vec![],
                    },
                    Some(usize::MAX),
                ),
        )
    }

    pub fn get(&self, id: NonZeroU32) -> Option<Vec<RowVal>> {
        // check wal first
        if let Some(val) = self.wal.get(id) {
            return Some(val);
        }

        // if not in pages, return None
        if self.pages.is_empty() {
            return None;
        }

        // otherwise, find the page where start <= id <= end
        let mut range = self.range_iter(id);

        match range.next() {
            Some(next_page) => next_page.0.get(id),
            None => None,
        }
    }

    pub fn remove(&mut self, id: NonZeroU32) -> Option<Vec<RowVal>> {
        // if in wal, remove from wal
        if let Some(val) = self.wal.remove(id) {
            return Some(val);
        }

        // if empty, return None
        if self.pages.is_empty() {
            return None;
        }

        // handle case when id is too small
        if let Some(first_page) = self.pages.first() {
            if id < first_page.0.header.start {
                return None;
            }
        }

        // handle case when id is too large
        if let Some(last_page) = self.pages.last() {
            if id > last_page.0.header.end {
                return None;
            }
        }

        // otherwise, find the page where start <= id <= end
        let mut range = self.range_iter(id);

        let next_page = range.next().unwrap();
        let mut fetched_page = next_page.clone();

        self.pages.remove(&fetched_page);
        let res = fetched_page.0.remove(id);

        // if the page still has items, readd it in
        if fetched_page.0.header.count != 0 {
            self.pages.insert(fetched_page);
        }

        res
    }

    pub fn insert(&mut self, id: NonZeroU32, val: &[RowVal]) {
        // if in wal, insert into wal
        if self.wal.insert(id, val) {
            return;
        }

        self.insert_to_page(id, val)
    }

    fn insert_to_page(&mut self, id: NonZeroU32, val: &[RowVal]) {
        let mut new_record = vec![RowVal::Id(id)];
        new_record.extend_from_slice(val);
        let row_size = val.iter().map(|x| x.size()).sum::<u16>() as usize + 4;

        // in case of an empty db
        if self.pages.is_empty() {
            let mut new_page = (Page::new_dirty(&[new_record], &self.schema.schema), None);
            new_page.0.size += row_size;
            self.pages.insert(new_page);
            return;
        }

        // handle prepend
        if let Some(first_page) = self.pages.first() {
            if id < first_page.0.header.start {
                let mut first_page = self.pages.pop_first().unwrap();
                first_page.0.size += row_size;
                first_page.0.insert(&new_record);
                self.pages.insert(first_page);

                // split page that is too big
                if let Some(first_page) = self.pages.first() {
                    if first_page.0.size() > PAGE_SIZE {
                        let (head, tail) = first_page.0.split();
                        self.pages.pop_first();
                        self.pages.insert((head, None));
                        self.pages.insert((tail, None));
                    }
                }
                return;
            }
        }

        // handle append
        if let Some(last_page) = self.pages.last() {
            if id > last_page.0.header.end {
                let mut last_page = self.pages.pop_last().unwrap();
                last_page.0.size += row_size;
                last_page.0.insert(&new_record);
                self.pages.insert(last_page);
                // split page that is too big
                if let Some(last_page) = self.pages.last() {
                    if last_page.0.size() > PAGE_SIZE {
                        let (head, tail) = last_page.0.split();
                        self.pages.pop_last();
                        self.pages.insert((head, None));
                        self.pages.insert((tail, None));
                    }
                }
                return;
            }
        }

        let mut range = self.range_iter(id);

        let next_page = range.next().unwrap();
        let mut fetched_page = next_page.clone();

        self.pages.remove(&fetched_page);
        fetched_page.0.insert(&new_record);

        if fetched_page.0.size() > PAGE_SIZE {
            let (head, tail) = fetched_page.0.split();
            self.pages.insert((head, None));
            self.pages.insert((tail, None));
        } else {
            self.pages.insert(fetched_page);
        }
    }
}

pub fn deserialize(bytes: Vec<u8>, schema: &[RowType]) -> BTreeSet<(Page, Option<usize>)> {
    assert!(bytes.len() % PAGE_SIZE == 0);

    let mut pages = vec![];

    for i in 0..(bytes.len() / PAGE_SIZE) {
        pages.push((
            Page::from_bytes(&bytes[i * PAGE_SIZE..(i + 1) * PAGE_SIZE], schema),
            Some(i),
        ));
    }

    BTreeSet::from_iter(pages)
}

impl Drop for DB {
    fn drop(&mut self) {
        self.serialize();
    }
}

#[cfg(test)]
mod tests {
    use std::{collections::HashMap, fs, num::NonZero};

    use insta::assert_yaml_snapshot as snapshot;

    use quickcheck_macros::quickcheck;

    use super::*;

    const DEFAULT_SCHEMA: &[RowType] = &[RowType::Id, RowType::U32];

    #[test]
    fn read_write() {
        let mut db = DB::new("tests/read_write", DEFAULT_SCHEMA);

        for i in 1..=5 {
            db.insert(NonZeroU32::new(i).unwrap(), &[RowVal::U32(i)]);
        }

        db.serialize();
        db.sync();

        let bytes = fs::read("tests/read_write.1.db").unwrap();

        let deserialized = deserialize(bytes, DEFAULT_SCHEMA);

        snapshot!(deserialized);
    }

    #[test]
    fn insert_loop() {
        let mut db = DB::new("tests/insert_loop", DEFAULT_SCHEMA);

        for i in 1..=510 {
            db.insert(NonZero::new(i).unwrap(), &[RowVal::U32(i)]);
        }

        db.sync();

        snapshot!(db.pages);
    }

    #[quickcheck]
    fn fuzz_db_get_insert(records: HashMap<NonZeroU32, u32>) -> bool {
        let mut db = DB::new("tests/fuzz_db_get", DEFAULT_SCHEMA);

        for (id, val) in &records {
            db.insert(*id, &[RowVal::U32(*val)]);
        }

        records
            .into_iter()
            .map(|(id, val)| db.get(id) == Some(vec![RowVal::U32(val)]))
            .all(|f| f)
    }
}
