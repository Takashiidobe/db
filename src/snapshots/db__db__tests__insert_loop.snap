---
source: src/db.rs
expression: db.pages
---
inner:
  - - - header:
          end: 510
          start: 1
          count: 510
        data:
          1:
            - U32: 1
          2:
            - U32: 2
          3:
            - U32: 3
          4:
            - U32: 4
          5:
            - U32: 5
          6:
            - U32: 6
          7:
            - U32: 7
          8:
            - U32: 8
          9:
            - U32: 9
          10:
            - U32: 10
          11:
            - U32: 11
          12:
            - U32: 12
          13:
            - U32: 13
          14:
            - U32: 14
          15:
            - U32: 15
          16:
            - U32: 16
          17:
            - U32: 17
          18:
            - U32: 18
          19:
            - U32: 19
          20:
            - U32: 20
          21:
            - U32: 21
          22:
            - U32: 22
          23:
            - U32: 23
          24:
            - U32: 24
          25:
            - U32: 25
          26:
            - U32: 26
          27:
            - U32: 27
          28:
            - U32: 28
          29:
            - U32: 29
          30:
            - U32: 30
          31:
            - U32: 31
          32:
            - U32: 32
          33:
            - U32: 33
          34:
            - U32: 34
          35:
            - U32: 35
          36:
            - U32: 36
          37:
            - U32: 37
          38:
            - U32: 38
          39:
            - U32: 39
          40:
            - U32: 40
          41:
            - U32: 41
          42:
            - U32: 42
          43:
            - U32: 43
          44:
            - U32: 44
          45:
            - U32: 45
          46:
            - U32: 46
          47:
            - U32: 47
          48:
            - U32: 48
          49:
            - U32: 49
          50:
            - U32: 50
          51:
            - U32: 51
          52:
            - U32: 52
          53:
            - U32: 53
          54:
            - U32: 54
          55:
            - U32: 55
          56:
            - U32: 56
          57:
            - U32: 57
          58:
            - U32: 58
          59:
            - U32: 59
          60:
            - U32: 60
          61:
            - U32: 61
          62:
            - U32: 62
          63:
            - U32: 63
          64:
            - U32: 64
          65:
            - U32: 65
          66:
            - U32: 66
          67:
            - U32: 67
          68:
            - U32: 68
          69:
            - U32: 69
          70:
            - U32: 70
          71:
            - U32: 71
          72:
            - U32: 72
          73:
            - U32: 73
          74:
            - U32: 74
          75:
            - U32: 75
          76:
            - U32: 76
          77:
            - U32: 77
          78:
            - U32: 78
          79:
            - U32: 79
          80:
            - U32: 80
          81:
            - U32: 81
          82:
            - U32: 82
          83:
            - U32: 83
          84:
            - U32: 84
          85:
            - U32: 85
          86:
            - U32: 86
          87:
            - U32: 87
          88:
            - U32: 88
          89:
            - U32: 89
          90:
            - U32: 90
          91:
            - U32: 91
          92:
            - U32: 92
          93:
            - U32: 93
          94:
            - U32: 94
          95:
            - U32: 95
          96:
            - U32: 96
          97:
            - U32: 97
          98:
            - U32: 98
          99:
            - U32: 99
          100:
            - U32: 100
          101:
            - U32: 101
          102:
            - U32: 102
          103:
            - U32: 103
          104:
            - U32: 104
          105:
            - U32: 105
          106:
            - U32: 106
          107:
            - U32: 107
          108:
            - U32: 108
          109:
            - U32: 109
          110:
            - U32: 110
          111:
            - U32: 111
          112:
            - U32: 112
          113:
            - U32: 113
          114:
            - U32: 114
          115:
            - U32: 115
          116:
            - U32: 116
          117:
            - U32: 117
          118:
            - U32: 118
          119:
            - U32: 119
          120:
            - U32: 120
          121:
            - U32: 121
          122:
            - U32: 122
          123:
            - U32: 123
          124:
            - U32: 124
          125:
            - U32: 125
          126:
            - U32: 126
          127:
            - U32: 127
          128:
            - U32: 128
          129:
            - U32: 129
          130:
            - U32: 130
          131:
            - U32: 131
          132:
            - U32: 132
          133:
            - U32: 133
          134:
            - U32: 134
          135:
            - U32: 135
          136:
            - U32: 136
          137:
            - U32: 137
          138:
            - U32: 138
          139:
            - U32: 139
          140:
            - U32: 140
          141:
            - U32: 141
          142:
            - U32: 142
          143:
            - U32: 143
          144:
            - U32: 144
          145:
            - U32: 145
          146:
            - U32: 146
          147:
            - U32: 147
          148:
            - U32: 148
          149:
            - U32: 149
          150:
            - U32: 150
          151:
            - U32: 151
          152:
            - U32: 152
          153:
            - U32: 153
          154:
            - U32: 154
          155:
            - U32: 155
          156:
            - U32: 156
          157:
            - U32: 157
          158:
            - U32: 158
          159:
            - U32: 159
          160:
            - U32: 160
          161:
            - U32: 161
          162:
            - U32: 162
          163:
            - U32: 163
          164:
            - U32: 164
          165:
            - U32: 165
          166:
            - U32: 166
          167:
            - U32: 167
          168:
            - U32: 168
          169:
            - U32: 169
          170:
            - U32: 170
          171:
            - U32: 171
          172:
            - U32: 172
          173:
            - U32: 173
          174:
            - U32: 174
          175:
            - U32: 175
          176:
            - U32: 176
          177:
            - U32: 177
          178:
            - U32: 178
          179:
            - U32: 179
          180:
            - U32: 180
          181:
            - U32: 181
          182:
            - U32: 182
          183:
            - U32: 183
          184:
            - U32: 184
          185:
            - U32: 185
          186:
            - U32: 186
          187:
            - U32: 187
          188:
            - U32: 188
          189:
            - U32: 189
          190:
            - U32: 190
          191:
            - U32: 191
          192:
            - U32: 192
          193:
            - U32: 193
          194:
            - U32: 194
          195:
            - U32: 195
          196:
            - U32: 196
          197:
            - U32: 197
          198:
            - U32: 198
          199:
            - U32: 199
          200:
            - U32: 200
          201:
            - U32: 201
          202:
            - U32: 202
          203:
            - U32: 203
          204:
            - U32: 204
          205:
            - U32: 205
          206:
            - U32: 206
          207:
            - U32: 207
          208:
            - U32: 208
          209:
            - U32: 209
          210:
            - U32: 210
          211:
            - U32: 211
          212:
            - U32: 212
          213:
            - U32: 213
          214:
            - U32: 214
          215:
            - U32: 215
          216:
            - U32: 216
          217:
            - U32: 217
          218:
            - U32: 218
          219:
            - U32: 219
          220:
            - U32: 220
          221:
            - U32: 221
          222:
            - U32: 222
          223:
            - U32: 223
          224:
            - U32: 224
          225:
            - U32: 225
          226:
            - U32: 226
          227:
            - U32: 227
          228:
            - U32: 228
          229:
            - U32: 229
          230:
            - U32: 230
          231:
            - U32: 231
          232:
            - U32: 232
          233:
            - U32: 233
          234:
            - U32: 234
          235:
            - U32: 235
          236:
            - U32: 236
          237:
            - U32: 237
          238:
            - U32: 238
          239:
            - U32: 239
          240:
            - U32: 240
          241:
            - U32: 241
          242:
            - U32: 242
          243:
            - U32: 243
          244:
            - U32: 244
          245:
            - U32: 245
          246:
            - U32: 246
          247:
            - U32: 247
          248:
            - U32: 248
          249:
            - U32: 249
          250:
            - U32: 250
          251:
            - U32: 251
          252:
            - U32: 252
          253:
            - U32: 253
          254:
            - U32: 254
          255:
            - U32: 255
          256:
            - U32: 256
          257:
            - U32: 257
          258:
            - U32: 258
          259:
            - U32: 259
          260:
            - U32: 260
          261:
            - U32: 261
          262:
            - U32: 262
          263:
            - U32: 263
          264:
            - U32: 264
          265:
            - U32: 265
          266:
            - U32: 266
          267:
            - U32: 267
          268:
            - U32: 268
          269:
            - U32: 269
          270:
            - U32: 270
          271:
            - U32: 271
          272:
            - U32: 272
          273:
            - U32: 273
          274:
            - U32: 274
          275:
            - U32: 275
          276:
            - U32: 276
          277:
            - U32: 277
          278:
            - U32: 278
          279:
            - U32: 279
          280:
            - U32: 280
          281:
            - U32: 281
          282:
            - U32: 282
          283:
            - U32: 283
          284:
            - U32: 284
          285:
            - U32: 285
          286:
            - U32: 286
          287:
            - U32: 287
          288:
            - U32: 288
          289:
            - U32: 289
          290:
            - U32: 290
          291:
            - U32: 291
          292:
            - U32: 292
          293:
            - U32: 293
          294:
            - U32: 294
          295:
            - U32: 295
          296:
            - U32: 296
          297:
            - U32: 297
          298:
            - U32: 298
          299:
            - U32: 299
          300:
            - U32: 300
          301:
            - U32: 301
          302:
            - U32: 302
          303:
            - U32: 303
          304:
            - U32: 304
          305:
            - U32: 305
          306:
            - U32: 306
          307:
            - U32: 307
          308:
            - U32: 308
          309:
            - U32: 309
          310:
            - U32: 310
          311:
            - U32: 311
          312:
            - U32: 312
          313:
            - U32: 313
          314:
            - U32: 314
          315:
            - U32: 315
          316:
            - U32: 316
          317:
            - U32: 317
          318:
            - U32: 318
          319:
            - U32: 319
          320:
            - U32: 320
          321:
            - U32: 321
          322:
            - U32: 322
          323:
            - U32: 323
          324:
            - U32: 324
          325:
            - U32: 325
          326:
            - U32: 326
          327:
            - U32: 327
          328:
            - U32: 328
          329:
            - U32: 329
          330:
            - U32: 330
          331:
            - U32: 331
          332:
            - U32: 332
          333:
            - U32: 333
          334:
            - U32: 334
          335:
            - U32: 335
          336:
            - U32: 336
          337:
            - U32: 337
          338:
            - U32: 338
          339:
            - U32: 339
          340:
            - U32: 340
          341:
            - U32: 341
          342:
            - U32: 342
          343:
            - U32: 343
          344:
            - U32: 344
          345:
            - U32: 345
          346:
            - U32: 346
          347:
            - U32: 347
          348:
            - U32: 348
          349:
            - U32: 349
          350:
            - U32: 350
          351:
            - U32: 351
          352:
            - U32: 352
          353:
            - U32: 353
          354:
            - U32: 354
          355:
            - U32: 355
          356:
            - U32: 356
          357:
            - U32: 357
          358:
            - U32: 358
          359:
            - U32: 359
          360:
            - U32: 360
          361:
            - U32: 361
          362:
            - U32: 362
          363:
            - U32: 363
          364:
            - U32: 364
          365:
            - U32: 365
          366:
            - U32: 366
          367:
            - U32: 367
          368:
            - U32: 368
          369:
            - U32: 369
          370:
            - U32: 370
          371:
            - U32: 371
          372:
            - U32: 372
          373:
            - U32: 373
          374:
            - U32: 374
          375:
            - U32: 375
          376:
            - U32: 376
          377:
            - U32: 377
          378:
            - U32: 378
          379:
            - U32: 379
          380:
            - U32: 380
          381:
            - U32: 381
          382:
            - U32: 382
          383:
            - U32: 383
          384:
            - U32: 384
          385:
            - U32: 385
          386:
            - U32: 386
          387:
            - U32: 387
          388:
            - U32: 388
          389:
            - U32: 389
          390:
            - U32: 390
          391:
            - U32: 391
          392:
            - U32: 392
          393:
            - U32: 393
          394:
            - U32: 394
          395:
            - U32: 395
          396:
            - U32: 396
          397:
            - U32: 397
          398:
            - U32: 398
          399:
            - U32: 399
          400:
            - U32: 400
          401:
            - U32: 401
          402:
            - U32: 402
          403:
            - U32: 403
          404:
            - U32: 404
          405:
            - U32: 405
          406:
            - U32: 406
          407:
            - U32: 407
          408:
            - U32: 408
          409:
            - U32: 409
          410:
            - U32: 410
          411:
            - U32: 411
          412:
            - U32: 412
          413:
            - U32: 413
          414:
            - U32: 414
          415:
            - U32: 415
          416:
            - U32: 416
          417:
            - U32: 417
          418:
            - U32: 418
          419:
            - U32: 419
          420:
            - U32: 420
          421:
            - U32: 421
          422:
            - U32: 422
          423:
            - U32: 423
          424:
            - U32: 424
          425:
            - U32: 425
          426:
            - U32: 426
          427:
            - U32: 427
          428:
            - U32: 428
          429:
            - U32: 429
          430:
            - U32: 430
          431:
            - U32: 431
          432:
            - U32: 432
          433:
            - U32: 433
          434:
            - U32: 434
          435:
            - U32: 435
          436:
            - U32: 436
          437:
            - U32: 437
          438:
            - U32: 438
          439:
            - U32: 439
          440:
            - U32: 440
          441:
            - U32: 441
          442:
            - U32: 442
          443:
            - U32: 443
          444:
            - U32: 444
          445:
            - U32: 445
          446:
            - U32: 446
          447:
            - U32: 447
          448:
            - U32: 448
          449:
            - U32: 449
          450:
            - U32: 450
          451:
            - U32: 451
          452:
            - U32: 452
          453:
            - U32: 453
          454:
            - U32: 454
          455:
            - U32: 455
          456:
            - U32: 456
          457:
            - U32: 457
          458:
            - U32: 458
          459:
            - U32: 459
          460:
            - U32: 460
          461:
            - U32: 461
          462:
            - U32: 462
          463:
            - U32: 463
          464:
            - U32: 464
          465:
            - U32: 465
          466:
            - U32: 466
          467:
            - U32: 467
          468:
            - U32: 468
          469:
            - U32: 469
          470:
            - U32: 470
          471:
            - U32: 471
          472:
            - U32: 472
          473:
            - U32: 473
          474:
            - U32: 474
          475:
            - U32: 475
          476:
            - U32: 476
          477:
            - U32: 477
          478:
            - U32: 478
          479:
            - U32: 479
          480:
            - U32: 480
          481:
            - U32: 481
          482:
            - U32: 482
          483:
            - U32: 483
          484:
            - U32: 484
          485:
            - U32: 485
          486:
            - U32: 486
          487:
            - U32: 487
          488:
            - U32: 488
          489:
            - U32: 489
          490:
            - U32: 490
          491:
            - U32: 491
          492:
            - U32: 492
          493:
            - U32: 493
          494:
            - U32: 494
          495:
            - U32: 495
          496:
            - U32: 496
          497:
            - U32: 497
          498:
            - U32: 498
          499:
            - U32: 499
          500:
            - U32: 500
          501:
            - U32: 501
          502:
            - U32: 502
          503:
            - U32: 503
          504:
            - U32: 504
          505:
            - U32: 505
          506:
            - U32: 506
          507:
            - U32: 507
          508:
            - U32: 508
          509:
            - U32: 509
          510:
            - U32: 510
        dirty: true
        size: 4088
        schema:
          - Id
          - U32
      - ~
index:
  inner:
    - 1
node_capacity: 1024
len: 1
